students/Lina/Lesson9/donor_models.py

```diff
@@ -0,0 +1,164 @@
+                 "donation will remain in your local community and be used by ChangeALife to provide programs",
+                 "and services to those in need.",
+                 "{new_paragraph_indent}On behalf of those whose lives will be impacted, please accept our sincere gratitude. May ",
+                 "you know the deep satisfaction of having made a difference in the lives of others.",
+                 "\n\n",
+                 "Below is a summary of your gift.",
+                 "",
+                 "Amount: ${donation_amount:.2f}",
+                 "Date: {date:%m-%d-%Y}",
+                 sender_template)).format(**letter_layout)
+        return letter
+
+    def letter_template(self, first_name, gift_amount):
+        #Creates a common format for some letters.
+        #Returns letter layout and sender template in a string
+        letter_layout = {"donor_first_name" : first_name,
+                         "donation_amount" : gift_amount,
+                         "sender_name" : "Tiffany Kurnett",
+                         "sender_title" : "President & CEO",
+                         "sender_alignment" : ' '*25,
+                         "new_paragraph_indent" : '\n   ',
+                         "date" : date.today()
+                        }
+        sender_template = "\n".join(("\n\n\n\n{sender_alignment}Sincerely,",
+                          "{sender_alignment}{sender_name}",
+                          "{sender_alignment}{sender_title}",
+                          "\n\n\n"))
+        return letter_layout, sender_template
+
+# DonorCollection class contains information about all donors
+class DonorCollection:
+    #This dictionary database contains the donor names and a history of their donation amounts.
+    def __init__(self):
+        self.donor_db = self.get_donors_history()
+
+    def donor_list(self):
+        #Returns a list with all donors sorted by name.
+        return sorted(self.donor_db.keys())
+
+    def lookup_donor(self, name):
+        #If donor is found, returns a dictionary with donor details
+        #If donor is not found, returns None
+        return self.donor_db.get(name)
+
+    def add_donation(self, name, amount):
+        #Adds a donation from a donor. If the donor does not exist, it will add to donor_db{} first.
+        self.donor_db.setdefault(name, {"gift_amounts":[]})
+        self.donor_db[name]["gift_amounts"].append(amount)
+
+    def donors_summary(self):
+        #Returns a list with summary of each donor in a tuple, sorted by total donated from
+        #highest to lowest.
+        # [(name, total donated, donation count, average amount)]
+        summary = []
+        for name, detail in self.donor_db.items():
+            d = Donor(name, detail)
+            summary.append((name, d.total_gift_amount, d.total_gift_count, d.avg_gift_amount))
```

Comment:
you should use the donor information as part of your collection, not create an instance on the fly - this defeats the purpose of having a Donor class.